src/main/scala/com/ebiznext/comet/job/ingest/IngestionJob.scala

```diff
@@ -931,6 +928,56 @@ trait IngestionJob extends SparkJob {
       finalSchema.map(attr => (attr.name, attr)).toMap
     dataFrame.columns.map(colName => attributesMap(colName)).toList
   }
+
+  private def updateBqTableSchema(table: Table, incomingSchema: StructType): Unit = {
+    val incomingFieldsMap = incomingSchema.fields.map(f => f.name -> f).toMap
+    val existingSchema =
+      table.getDefinition.asInstanceOf[StandardTableDefinition].getSchema.getFields()
+    val existingFieldNames = existingSchema.asScala.map(_.getName).toList
+    val newFields = incomingFieldsMap.keys.filterNot(existingFieldNames.toSet)
+    val oldFields = existingFieldNames.filterNot(incomingFieldsMap.keys.toSet)
+    if (oldFields.nonEmpty)
+      throw new Exception(
+        s"existing table contains fields not present in the new schema:  $oldFields"
+      )
+
+    def updateSchema() = {
+      val newBqSchema = BigQueryUtils.bqSchema(incomingSchema)
+      val updatedTable =
+        table.toBuilder.setDefinition(StandardTableDefinition.of(newBqSchema)).build()
+      updatedTable.update()
+    }
+
+    if (allowFieldAddition && newFields.nonEmpty) {
+      logger.info(s"The following new columns have been detected: ${newFields.toList}")
+      newFields.foreach { fieldName => // make sure all new fields are nullable
+        val field = incomingFieldsMap(fieldName)
+        if (!field.nullable) {
+          throw new Exception(s"Cannot add required $field with name $fieldName")
+        }
+      }
+      updateSchema()
+    }
+    if (allowFieldRelaxation) {
+      val existingFields = existingSchema.asScala
+        .map(f => (f.getName, Option(f.getMode).forall(_ == Field.Mode.NULLABLE)))
+        .toMap
+      val incomingRequiredFields = incomingSchema.fields.filter(!_.nullable).map(_.name)
+      incomingRequiredFields.foreach { newFieldName =>
+        val existingField = existingFields.get(newFieldName)
+        existingField match {
+          case None => logger.info("This is a new field, it is handled by allowFieldAddition Above")
+          case Some(true) =>
+            throw new Exception(
+              s"Merge failed because existing field is nullable but incoming one is required -> $newFieldName"
+            )
+          case Some(false) => // unchanged field
+        }
+      }
+      if (!allowFieldAddition || newFields.isEmpty)
+        updateSchema()
+    }
+  }
```

Comment:
This algorithm does not handle records. You should have a look at MergeUtils#computeCompatibleSchema, it almost does the same thing in the end : if the produced StructType is different from the existing one, then an update is required.
```suggestion
  private def updateBqTableSchema(table: Table, incomingSchema: StructType): Unit = {
    val existingSchema = BigQuerySchemaConverters.toSpark(
      table.getDefinition.asInstanceOf[StandardTableDefinition].getSchema
    )

    val compatibleSchema = MergeUtils.computeCompatibleSchema(existingSchema, incomingSchema)
    if ((allowFieldAddition || allowFieldRelaxation) && compatibleSchema != incomingSchema) {
      val newBqSchema = BigQueryUtils.bqSchema(incomingSchema)
      val updatedTable =
        table.toBuilder.setDefinition(StandardTableDefinition.of(newBqSchema)).build()
      updatedTable.update()
    }
  }
```

Reply:
Actually we do not need all these checks since BQ will throw an exception if schemas are incompatible